### DISCUSSION TITLE
Support or example for Wonderware system platform

### DISCUSSION THREAD

**Discussion by MattsPurcell** [FreeOpcUa/opcua-asyncio#1255] [Q&A]
Has anyone attempted to configure an OPC UA server from this code repository for use with Wonderware system platform using the OI gateway driver?

**Comment by MattsPurcell** [comment]
Background: I'm working with a new robotics vendor. They have a Linux edge server and have configured it as an OPC UA server. I need to integrate this into our Plant SCADA, however our SCADA support vendor says their server implementation does not behave like a proper OPC UA server. Without getting deep into the weeds, I'm curious if anyone in the community has been down this path.

**Reply by AndreasHeine** [reply]
interestingly it works with most other... we need to get more details whats not working like what they expect and then we can compare against opcua spec.

**Reply by MattsPurcell** [reply]
Sure. I can email this directly if more helpful also. Using the wonderware OI gateway driver, we are able to see and connect to the OPC UA server running on the linux machine. I hope the Screen shots are visible, I can also paste them as attachments if necessary
The namespace is populated automatically
<img width="455" alt="image" src="https://user-images.githubusercontent.com/128642642/227223698-61ee2de9-0c70-498c-a1ff-714f375bb6b7.png">

After connection is established, we expect to be able to view the tags in the built in OPC tag browser. All of the relevant information should be contained in the structure get_dorainductor_status_parent , however the tag browser is blank, no tags visible. Our wonderware vendor tells us that this is not behaving like a proper OPC UA server
<img width="581" alt="image" src="https://user-images.githubusercontent.com/128642642/227224325-d3f5f4b5-83cc-4f9b-8ada-eef8c86a48db.png">

I will paste the code for the server instance provided to us by the vendor below

**Reply by MattsPurcell** [reply]
import asyncio
import time
from typing import Dict

from asyncua import ua, Server, uamethod
from asyncua.common.structures104 import new_struct, new_struct_field


import json
import datetime
from pathlib import Path
import requests
from urllib3.exceptions import (
    InsecureRequestWarning,
)
from urllib3 import (
    disable_warnings,
)

from scada_gateway.custom_logger import get_logger


disable_warnings(InsecureRequestWarning)

from scada_gateway.gateway_types import Config, WORKCELL_STATUS_MAP


class ScadaGateway:
    def __init__(self, config: Config, request_cert_file: str = False) -> None:
        # self.logger = logging.getLogger("scada-gateway")
        self.logger = get_logger("scada-gateway")
        # read config
        self.config = config

        self.reqeust_cert_file = request_cert_file
        # init web_ui_server url
        c_ = self.config.web_ui_server
        self.setup_info_url = f"https://{c_.host}:{c_.port}/{c_.api.setup}"

    def wait_until_web_ui_server_is_ready(self, interval_s: float):
        c_ = self.config.web_ui_server
        ping_url = f"https://{c_.host}:{c_.port}"
        self.logger.info(f"waiting for web_ui_server to be ready: {ping_url}")
        while True:
            try:
                requests.get(
                    ping_url,
                    verify=self.reqeust_cert_file,
                    timeout=10.0,
                )
            except (requests.ConnectionError, requests.ConnectTimeout) as e:
                self.logger.warning("web_ui_server is not ready. wait again ... ")
            except Exception as e:
                self.logger.warning(
                    "error found while ping web_ui_server. wait again ... "
                )
                self.logger.error(f"{e}")
            else:
                break
            finally:
                time.sleep(interval_s)

    async def init_opcua_server(self):
        self.server = Server()
        await self.server.init()

        c_ = self.config.opcua_server
        url = f"opc.tcp://{c_.host}:{c_.port}"

        self.logger.info(f"starting OPC-UA server at {url}")
        self.server.set_endpoint(url)

        # setup our own namespace, not really necessary but should as spec
        self.logger.info(f"OPC-UA namesapce {c_.uri}")
        idx = await self.server.register_namespace(c_.uri)

        sn1, _ = await new_struct(
            self.server,
            idx,
            "SystemNotification",
            [
                new_struct_field("errorType", ua.VariantType.String),
                new_struct_field("timestamp", ua.VariantType.Int64),
                new_struct_field("errorCode", ua.VariantType.String),
                new_struct_field("component", ua.VariantType.String),
                new_struct_field("description", ua.VariantType.String),
            ],
        )

        sn2, _ = await new_struct(
            self.server,
            idx,
            "DoraInductorStatus",
            [
                new_struct_field("success", ua.VariantType.Boolean),
                new_struct_field("timestamp", ua.VariantType.Int64),
                new_struct_field("systemStatus", ua.VariantType.String),
                new_struct_field("currentShiftPackageCount", ua.VariantType.Int32),
                new_struct_field("throughput", ua.VariantType.Int32),
                new_struct_field("systemNotifications", sn1, array=True),
            ],
        )

        await self.server.load_data_type_definitions()

        objects = self.server.nodes.objects

        obj = await objects.add_object(idx, "get_dorainductor_status_parent")

        out_arg = ua.Argument()
        out_arg.Name = "doraInductorStatus"
        out_arg.DataType = sn2
        out_arg.ValueRank = -1
        out_arg.ArrayDimensions = []
        out_arg.Description = ua.LocalizedText("arg indicator")

        @uamethod
        async def callback(parent):
            self.logger.info(f"{parent}")
            self.logger.info("enter callback ")
            res = await self.get_setup_status()

            self.logger.info(f"response:\n {json.dumps(res, indent=2)}")

            notifications = res["systemNotifications"]
            del res["systemNotifications"]

            ret = (
                ua.Variant(
                    ua.DoraInductorStatus(
                        **res,
                        systemNotifications=list(
                            map(
                                lambda x: ua.SystemNotification(**x),
                                notifications,
                            )
                        ),
                    ),
                    ua.VariantType.ExtensionObject,
                ),
            )
            return ret

        await obj.add_method(
            idx,
            "get_dorainductor_status",
            callback,
            [out_arg],
        )

    async def get_setup_status(self):
        # default reponse
        res = {
            "success": False,
            "systemStatus": "",
            "currentShiftPackageCount": -1,
            "throughput": -1,
            "timestamp": int(datetime.datetime.now().timestamp() * 1000),
            "systemNotifications": [],
        }

        try:
            response = requests.post(
                self.setup_info_url,
                json={"workcell_id": self.config.workcell_id},
                verify=self.reqeust_cert_file,
                timeout=10.0,
            )
            _res = {}
            self.logger.info("received _response from web_ui_server")
            if response.status_code == 200 and response.json()["code"] == 0:
                self.logger.error("successfully received data. ")
                payload = response.json()["payload"]
                # self.logger.info(f"{payload}")
                status = payload["setup_status"]
                _res["systemStatus"] = WORKCELL_STATUS_MAP[status["workcell_status"]]
                _res["currentShiftPackageCount"] = int(status["package_count"])
                _res["throughput"] = int(status["pph"])

                # process system notifications
                _res["systemNotifications"] = list(
                    map(
                        lambda x: {
                            "errorType": "emergency",
                            "timestamp": int(x["emergency_time"]),
                            "errorCode": x["emergency_content"]["error_id"],
                            "component": x["emergency_content"]["component_id"],
                            "description": x["emergency_content"]["error_code"],
                        },
                        status["emergency_list"],
                    )
                ) + list(
                    map(
                        lambda x: {
                            "errorType": "warning",
                            "timestamp": int(x["warning_time"]),
                            "errorCode": x["warning_content"]["error_id"],
                            "component": x["warning_content"]["component_id"],
                            "description": x["warning_content"]["error_code"],
                        },
                        status["warning_list"],
                    )
                )
                _res["systemNotifications"].sort(key=lambda x: x["timestamp"])

                _res["success"] = True

                res = _res
            else:
                self.logger.error(f"{response.json() }")
                self.logger.error("web_ui_server can not send back useful data. ")

        except (requests.ConnectionError, requests.ConnectTimeout) as e:
            self.logger.error(
                f"Connection failed, please make sure the web_ui_server is alive and listen to port {self.setup_info_url}. "
            )
        except Exception as e:
            self.logger.error(f"Other exception found. Should crash the server. ")
            self.logger.error(f"{e}")

        finally:
            return res

    def start(self):
        self.logger.info("starting the opcua server. ")

        # self.wait_until_web_ui_server_is_ready(interval_s=1.0)

        async def func():
            await self.init_opcua_server()
            async with self.server:
                while True:
                    await asyncio.sleep(1)

        asyncio.run(func())


def main():
    CONFIG_PATH = Path(__file__).parent.parent / "config.json"
    # REQUEST_CERT_PATH = Path(__file__).parent.parent / "security" / "crt.pem"
    ScadaGateway(Config(**json.loads(CONFIG_PATH.read_text()))).start()


if __name__ == "__main__":
    main()

**Reply by MattsPurcell** [reply]
I'm curious if there is a problem with the way the tag structure is generated. Please let me know if I can share anything else for clarity

**Reply by MattsPurcell** [reply]
And thank you in advance!

**Reply by AndreasHeine** [reply]
does it work with UaExpert? its the OPCUA reference Client tool!

**Reply by schroeder-** [reply]
What are you expecting. From a quick browse you should only find 1 method get_dorainductor_status? Maybe they don't support methods and method calls?

**Reply by MattsPurcell** [reply]
The expectation, as I understand, is that these tags in these structures should be displayed in a granular hierarchy in the tag browser
"SystemNotification",
        [
            new_struct_field("errorType", ua.VariantType.String),
            new_struct_field("timestamp", ua.VariantType.Int64),
            new_struct_field("errorCode", ua.VariantType.String),
            new_struct_field("component", ua.VariantType.String),
            new_struct_field("description", ua.VariantType.String),
        ],
    )

    sn2, _ = await new_struct(
        self.server,
        idx,
        "DoraInductorStatus",
        [
            new_struct_field("success", ua.VariantType.Boolean),
            new_struct_field("timestamp", ua.VariantType.Int64),
            new_struct_field("systemStatus", ua.VariantType.String),
            new_struct_field("currentShiftPackageCount", ua.VariantType.Int32),
            new_struct_field("throughput", ua.VariantType.Int32),
            new_struct_field("systemNotifications", sn1, array=True),
        ],
    )



That certainly doesn't mean that the tags are not available to our SCADA, but it's really throwing us off. 

We address tags in the following manner    'ServerName'.'TopicName'.'TagName'

I have the server and topic names, but I'm unclear about the tag name. would it be /get_dorainductor_status_parent/SsytemNotification/timestamp    for example?

**Reply by AndreasHeine** [reply]
we dont support the ExposeItsArrayFeature...

thats actually not an mandatory feature of opcua!

if a server support that it would be visible under Server -> ServerCapabilities -> ModellingRules -> ExposesItsArray

**Reply by AndreasHeine** [reply]
so with that in mind Wonderware is wrong... If you want to have the nodes visible you need to add them to your server addresspace as childnodes of the structure variable and connect them to the python variables (so they have the same value)

**Reply by MattsPurcell** [reply]
I have no control over the OPC server, only the wonderware side. So no issues from me if the arrays are not visible, it's a small application with less than 20 tags, I just need some guidance on how to properly configure the tags to pull in the data please.

**Reply by AndreasHeine** [reply]
does Wonderware support custom Structures/ExtensionObjects? If so they should have a Manual for that!

**Reply by schroeder-** [reply]
I think you mix somethings up. new_struct is just a custom datatype. They are only a blueprint. If you want a tag a client can write/read you have to create a variable. Or if you want to group something you have to create a object with variables as child.

**Reply by MattsPurcell** [reply]
Thanks for your help today everyone, I am circling back with Wonderware for clarification on their driver
